File: minimum_size_subarray_sum.cpp
/*
209. Minimum Size Subarray Sum

Time Complexity: O(n)
Space Complexity: O(1)
*/

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;


class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int n = nums.size();
        int minLength = INT_MAX;
        int windowSum = 0;
        int left = 0;
        
        for (int right = 0; right < n; right++) {
            windowSum += nums[right];
            while (windowSum >= target) {
                minLength = min(minLength, right - left + 1);
                windowSum -= nums[left];
                left++;
            }
        }
        
        return minLength == INT_MAX ? 0 : minLength;
    }
};


// Test cases
int main() {
    Solution solution;
    
    // Example 1
    vector<int> nums1 = {2, 3, 1, 2, 4, 3};
    cout << "Example 1: " << solution.minSubArrayLen(7, nums1) << endl;  // Expected: 2
    // Subarray [4, 3] has sum 7
    
    // Example 2
    vector<int> nums2 = {1, 4, 4};
    cout << "Example 2: " << solution.minSubArrayLen(4, nums2) << endl;  // Expected: 1
    // Single element [4] has sum 4
    
    // Example 3
    vector<int> nums3 = {1, 1, 1, 1, 1, 1, 1, 1};
    cout << "Example 3: " << solution.minSubArrayLen(11, nums3) << endl;  // Expected: 0
    // Maximum sum is 8, cannot reach 11
    
    // Edge case: single element matching
    vector<int> nums4 = {5};
    cout << "Example 4: " << solution.minSubArrayLen(5, nums4) << endl;  // Expected: 1
    
    // Edge case: entire array needed
    vector<int> nums5 = {1, 2, 3, 4, 5};
    cout << "Example 5: " << solution.minSubArrayLen(15, nums5) << endl;  // Expected: 5
    // Sum = 15
    
    // Edge case: first element is sufficient
    vector<int> nums6 = {5, 1, 1, 1};
    cout << "Example 6: " << solution.minSubArrayLen(3, nums6) << endl;  // Expected: 1
    
    // Edge case: target = 1 (minimum)
    vector<int> nums7 = {1, 2, 3};
    cout << "Example 7: " << solution.minSubArrayLen(1, nums7) << endl;  // Expected: 1
    
    // Edge case: large target, cannot achieve
    vector<int> nums8 = {1, 2, 3, 4, 5};
    cout << "Example 8: " << solution.minSubArrayLen(100, nums8) << endl;  // Expected: 0
    
    // Edge case: consecutive elements
    vector<int> nums9 = {1, 2, 3, 4, 5, 6, 7, 8};
    cout << "Example 9: " << solution.minSubArrayLen(15, nums9) << endl;  // Expected: 5
        
    return 0;
}

File: minimum_size_subarray_sum.py
"""
209. Minimum Size Subarray Sum

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        min_length = float('inf')
        window_sum = 0
        left = 0
        
        for right in range(n):
            window_sum += nums[right]
            while window_sum >= target:
                min_length = min(min_length, right - left + 1)
                window_sum -= nums[left]
                left += 1
        
        return min_length if min_length != float('inf') else 0


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    target1 = 7
    nums1 = [2, 3, 1, 2, 4, 3]
    print(f"Example 1: {solution.minSubArrayLen(target1, nums1)}")  # Expected: 2
    # Subarray [4, 3] has sum 7
    
    # Example 2
    target2 = 4
    nums2 = [1, 4, 4]
    print(f"Example 2: {solution.minSubArrayLen(target2, nums2)}")  # Expected: 1
    # Single element [4] has sum 4
    
    # Example 3
    target3 = 11
    nums3 = [1, 1, 1, 1, 1, 1, 1, 1]
    print(f"Example 3: {solution.minSubArrayLen(target3, nums3)}")  # Expected: 0
    # Maximum sum is 8, cannot reach 11
    
    # Edge case: single element matching
    target4 = 5
    nums4 = [5]
    print(f"Example 4: {solution.minSubArrayLen(target4, nums4)}")  # Expected: 1
    
    # Edge case: entire array needed
    target5 = 15
    nums5 = [1, 2, 3, 4, 5]
    print(f"Example 5: {solution.minSubArrayLen(target5, nums5)}")  # Expected: 5
    # Sum = 15
    
    # Edge case: first element is sufficient
    target6 = 3
    nums6 = [5, 1, 1, 1]
    print(f"Example 6: {solution.minSubArrayLen(target6, nums6)}")  # Expected: 1
    
    # Edge case: target = 1 (minimum)
    target7 = 1
    nums7 = [1, 2, 3]
    print(f"Example 7: {solution.minSubArrayLen(target7, nums7)}")  # Expected: 1
    
    # Edge case: large target, cannot achieve
    target8 = 100
    nums8 = [1, 2, 3, 4, 5]
    print(f"Example 8: {solution.minSubArrayLen(target8, nums8)}")  # Expected: 0
    
    # Edge case: consecutive elements
    target9 = 15
    nums9 = [1, 2, 3, 4, 5, 6, 7, 8]
    print(f"Example 9: {solution.minSubArrayLen(target9, nums9)}")  # Expected: 5
    # [1,2,3,4,5] = 15 or [2,3,4,6] = 15 (but need to check which is shorter)
